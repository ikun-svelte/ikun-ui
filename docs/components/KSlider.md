---
title: KSlider
lang: en-US
---

# KSlider

Drag the slider within a fixed range.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/slider
```

```bash [yarn]
yarn add @ikun-ui/slider
```

```bash [npm]
npm install @ikun-ui/slider
```

:::

## Basic usage

<demo src="slider/basic.svelte" github="Slider"></demo>

## Disabled slider

Use `disabled` attribute to determine whether a slider is disabled.

<demo src="slider/disabled.svelte" github="Slider"></demo>

## Maximum and minimum values

The `max` and `min` attributes limit the maximum and minimum values of the `slider`

<demo src="slider/range.svelte" github="Slider"></demo>

## Custom Render slider

Rendering `slider` content using slots

<demo src="slider/render.svelte" github="Slider"></demo>

## Sizes

Add `size` attribute to change the size of Slider. It supports `sm`, `md` and `lg`.

<demo src="slider/sizes.svelte" github="Slider"></demo>

## Slider Props

| Name     | Type                     | Default | Description                |
| -------- | ------------------------ | ------- | -------------------------- |
| value    | `number`                 | `0`     | Binding value              |
| min      | `number`                 | `0`     | Minimum value              |
| max      | `number`                 | `100`   | Maximum value              |
| step     | `number`                 | `1`     | Step size                  |
| disabled | `boolean`                | `false` | Whether Slider is disabled |
| size     | `sm \| md \| lg`         | `md`    | Size of `slider`.          |
| cls      | `string`                 | `-`     | Additional class           |
| attrs    | `Record<string, string>` | `{}`    | Additional attributes      |

## Slider Events

| Name   | Description                                            | Type                          |
| ------ | ------------------------------------------------------ | ----------------------------- |
| change | Fires when the value changes and the mouse is released | `(value: CustomEvent)=> void` |
| input  | Trigger in real time when data changes                 | `(value: CustomEvent)=> void` |

## Slider Slots

| Name         | Description                     |
| ------------ | ------------------------------- |
| buttonRender | Customize slider button content |
