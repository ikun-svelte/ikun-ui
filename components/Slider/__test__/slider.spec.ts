import KSlider from '../src';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { tick } from 'svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KSlider', () => {
	test('props: value', async () => {
		const value = 10;
		const instance = new KSlider({
			target: host,
			props: {
				value
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const sliderBarElm = host.getElementsByClassName('k-slider--bar')[0] as HTMLDivElement;
		const sliderButtonWrapperElm = host.getElementsByClassName(
			'k-slider--button__wrapper'
		)[0] as HTMLDivElement;
		expect(Number.parseFloat(sliderBarElm.style.width)).toBe(value);
		expect(Number.parseFloat(sliderButtonWrapperElm.style.left)).toBe(value);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: min and max', async () => {
		const value = 1;
		const instance = new KSlider({
			target: host,
			props: {
				value,
				min: 10,
				max: 20
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const sliderBarElm = host.getElementsByClassName('k-slider--bar')[0] as HTMLDivElement;
		const sliderButtonWrapperElm = host.getElementsByClassName(
			'k-slider--button__wrapper'
		)[0] as HTMLDivElement;
		expect(Number.parseFloat(sliderBarElm.style.width)).toBe(0);
		expect(Number.parseFloat(sliderButtonWrapperElm.style.left)).toBe(0);
		instance.$set({
			value: 15
		});
		await tick();
		expect(Number.parseFloat(sliderBarElm.style.width)).toBe(50);
		expect(Number.parseFloat(sliderButtonWrapperElm.style.left)).toBe(50);
		instance.$set({
			value: 30
		});
		await tick();
		expect(Number.parseFloat(sliderBarElm.style.width)).toBe(100);
		expect(Number.parseFloat(sliderButtonWrapperElm.style.left)).toBe(100);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		const instance = new KSlider({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('k-cur-disabled')).toBe(true);
		instance.$set({
			disabled: false
		});
		await tick();
		expect(host.innerHTML.includes('k-cur-disabled')).toBe(false);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: size', async () => {
		const instanceSM = new KSlider({
			target: host,
			props: {
				size: 'sm'
			}
		});
		const instanceMD = new KSlider({
			target: host,
			props: {
				size: 'md'
			}
		});
		const instanceLG = new KSlider({
			target: host,
			props: {
				size: 'lg'
			}
		});
		expect(instanceSM).toBeTruthy();
		expect(instanceMD).toBeTruthy();
		expect(instanceLG).toBeTruthy();
		await tick();

		expect(host.innerHTML.includes('k-slider--button__sm')).toBe(true);
		expect(host.innerHTML.includes('k-slider--button__md')).toBe(true);
		expect(host.innerHTML.includes('k-slider--button__lg')).toBe(true);
		expect(host.innerHTML.includes('k-slider--button__wrapper__sm')).toBe(true);
		expect(host.innerHTML.includes('k-slider--button__wrapper__md')).toBe(true);
		expect(host.innerHTML.includes('k-slider--button__wrapper__lg')).toBe(true);
		expect(host.innerHTML.includes('k-slider__sm')).toBe(true);
		expect(host.innerHTML.includes('k-slider__md')).toBe(true);
		expect(host.innerHTML.includes('k-slider__lg')).toBe(true);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: cls', async () => {
		const instance = new KSlider({
			target: host,
			props: {
				cls: '窈窕淑女，君子好逑'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('窈窕淑女，君子好逑')).toBe(true);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: attrs', async () => {
		const instance = new KSlider({
			target: host,
			props: {
				attrs: {
					you: 'world'
				}
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const sliderBaseElm = host.getElementsByClassName('k-slider--base')[0] as HTMLDivElement;
		expect(sliderBaseElm.getAttribute('you')).toBe('world');
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: should trigger change event', async () => {
		const changeEvent = vi.fn();
		const instance = new KSlider({
			target: host,
			props: {
				value: 0
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('change', () => {
			changeEvent();
		});
		const sliderButtonWrapperElm = host.getElementsByClassName(
			' k-slider--button__wrapper'
		)[0] as HTMLDivElement;
		sliderButtonWrapperElm.dispatchEvent(new Event('mousedown'));
		window.dispatchEvent(new Event('mouseup'));
		await tick();
		expect(changeEvent).toBeCalled();
	});

	test('event: should trigger input event', async () => {
		const inputEvent = vi.fn();
		const instance = new KSlider({
			target: host,
			props: {
				value: 0
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('input', () => {
			inputEvent();
		});
		const sliderRunwayElm = host.getElementsByClassName(' k-slider--runway')[0] as HTMLDivElement;
		sliderRunwayElm.dispatchEvent(
			new MouseEvent('mousedown', {
				cancelable: true,
				clientX: 100
			})
		);
		await tick();
		expect(inputEvent).toBeCalled();
	});

	test('event: should not trigger event when disabled', async () => {
		const event = vi.fn();
		const instance = new KSlider({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('change', event);
		instance.$on('input', event);
		const sliderButtonWrapperElm = host.getElementsByClassName(
			' k-slider--button__wrapper'
		)[0] as HTMLDivElement;
		sliderButtonWrapperElm.dispatchEvent(new Event('mousedown'));
		window.dispatchEvent(new Event('mouseup'));
		const sliderRunwayElm = host.getElementsByClassName(' k-slider--runway')[0] as HTMLDivElement;
		sliderRunwayElm.dispatchEvent(
			new MouseEvent('mousedown', {
				cancelable: true,
				clientX: 100
			})
		);
		await tick();
		expect(event).not.toBeCalled();
	});
});
