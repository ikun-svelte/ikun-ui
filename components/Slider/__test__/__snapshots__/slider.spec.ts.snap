// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KSlider > props: attrs 1`] = `"<div class=\\"k-slider k-slider--base k-slider--md\\" you=\\"world\\"><div class=\\"k-slider--runway--md k-slider--runway\\" aria-hidden=\\"true\\"><div class=\\"k-slider--button-wrapper\\" aria-hidden=\\"true\\" style=\\"left: 0%\\"><div class=\\"k-slider--button k-slider--button--md\\"></div></div> <div class=\\"k-slider--bar\\" style=\\"width: 0%; left: 0%\\"></div></div></div>"`;

exports[`Test: KSlider > props: cls 1`] = `"<div class=\\"k-slider k-slider--base k-slider--md 窈窕淑女，君子好逑\\"><div class=\\"k-slider--runway--md k-slider--runway\\" aria-hidden=\\"true\\"><div class=\\"k-slider--button-wrapper\\" aria-hidden=\\"true\\" style=\\"left: 0%\\"><div class=\\"k-slider--button k-slider--button--md\\"></div></div> <div class=\\"k-slider--bar\\" style=\\"width: 0%; left: 0%\\"></div></div></div>"`;

exports[`Test: KSlider > props: disabled 1`] = `"<div class=\\"k-slider k-slider--base k-slider--md\\"><div class=\\"k-slider--runway--md k-slider--runway\\" aria-hidden=\\"true\\"><div class=\\"k-slider--button-wrapper\\" aria-hidden=\\"true\\" style=\\"left: 0%\\"><div class=\\"k-slider--button k-slider--button--md\\"></div></div> <div class=\\"k-slider--bar\\" style=\\"width: 0%; left: 0%\\"></div></div></div>"`;

exports[`Test: KSlider > props: min and max 1`] = `"<div class=\\"k-slider k-slider--base k-slider--md\\"><div class=\\"k-slider--runway--md k-slider--runway\\" aria-hidden=\\"true\\"><div class=\\"k-slider--button-wrapper\\" aria-hidden=\\"true\\" style=\\"left: 100%\\"><div class=\\"k-slider--button k-slider--button--md\\"></div></div> <div class=\\"k-slider--bar\\" style=\\"width: 100%; left: 0%\\"></div></div></div>"`;

exports[`Test: KSlider > props: size 1`] = `"<div class=\\"k-slider k-slider--base k-slider--lg\\"><div class=\\"k-slider--runway--lg k-slider--runway\\" aria-hidden=\\"true\\"><div class=\\"k-slider--button-wrapper\\" aria-hidden=\\"true\\" style=\\"left: 0%\\"><div class=\\"k-slider--button k-slider--button--lg\\"></div></div> <div class=\\"k-slider--bar\\" style=\\"width: 0%; left: 0%\\"></div></div></div>"`;

exports[`Test: KSlider > props: value 1`] = `"<div class=\\"k-slider k-slider--base k-slider--md\\"><div class=\\"k-slider--runway--md k-slider--runway\\" aria-hidden=\\"true\\"><div class=\\"k-slider--button-wrapper\\" aria-hidden=\\"true\\" style=\\"left: 10%\\"><div class=\\"k-slider--button k-slider--button--md\\"></div></div> <div class=\\"k-slider--bar\\" style=\\"width: 10%; left: 0%\\"></div></div></div>"`;

exports[`Test: KSlider > props: vertical 1`] = `"<div class=\\"k-slider k-slider--base__vertical k-slider--md__vertical\\"><div class=\\"k-slider--runway--md__vertical k-slider--runway__vertical\\" aria-hidden=\\"true\\"><div class=\\"k-slider--button-wrapper__vertical\\" aria-hidden=\\"true\\" style=\\"bottom: 0%\\"><div class=\\"k-slider--button k-slider--button--md\\"></div></div> <div class=\\"k-slider--bar__vertical\\" style=\\"height: 0%; bottom: 0%\\"></div></div></div>"`;
